index.js

```diff
@@ -47,10 +42,85 @@ module.exports = {
   },
 
   forEach: function(path, cb, thisArg) {
-    forEach(path.match(SPLIT_REGEX), cb, thisArg)
+    forEach(split(path), cb, thisArg)
+  }
+}
+
+function setterEval(path) {
+  return (
+    setCache[path] ||
+    (setCache[path] = new Function(
+      'data, value',
+      expr(path, 'data') + ' = value'
+    ))
+  )
+}
+
+function setterFallbackMemoized(path) {
+  var parts = normalizePath(path)
+  return function(data, value) {
+    return setterFallback(parts, data, value)
+  }
+}
+
+function setterFallback(parts, data, value) {
+  var index = 0,
+    len = parts.length
+  while (index < len - 1) {
+    data = data[parts[index++]]
+  }
+  data[parts[index]] = value
+}
+
+function getterEval(path, safe) {
+  var k = path + '_' + safe
+  return (
+    getCache[k] ||
+    (getCache[k] = new Function('data', 'return ' + expr(path, safe, 'data')))
+  )
+}
+
+function getterFallbackMemoized(path, safe) {
+  var parts = normalizePath(path)
+  return function(data) {
+    return getterFallback(parts, safe, data)
   }
 }
 
+function getterFallback(parts, safe, data) {
+  var index = 0,
+    len = parts.length
+  while (index < len) {
+    if (data || !safe) {
+      data = data[parts[index++]]
+    } else {
+      return
+    }
+  }
+  return data
+}
+
+function normalizePath(path) {
+  var parts = pathCache[path]
+  if (parts) {
+    return parts
+  }
+  if (pathCacheSize >= MAX_CACHE_SIZE) {
+    pathCache = {}
+    pathCacheSize = 1
```

Comment:
lets wrap this up into a little reusable function use it for the other caches as well

```js
const cache = () => {
  let size = 1;
  let values = {};

  return {
    get: key => values[key],
    set: (key, value) => {
      if (size >= MAX_CACHE_SIZE) {
        values = {};
        size = 1;
      } else {
        size += 1;
      }
      return (values[key] = value);
    },
  };
};
```